saige_assoc.py

```diff
@@ -0,0 +1,383 @@
+    plink_path: str,  # ./input/nfam_100_nindep_0_step1_includeMoreRareVariants_poly
+    pheno_col: str = 'y',  # ${gene_name}, can be random y because we are doing one gene
+    trait_type: str = 'count',
+    # this is a boolean but that's encoded differently between R and python
+    skip_vre: str = 'FALSE',
+    pheno_remove_zeros: str = 'FALSE',
+    use_sparse_grm_null: str = 'FALSE',
+    use_grm_null: str = 'FALSE',
+    is_cov_offset: str = 'FALSE',
+    is_cov_transform: str = 'TRUE',
+    skip_model_fitting: str = 'FALSE',
+    tol: float = 0.00001,
+    is_overwrite_vre_file: str = 'TRUE',
+):
+    """Build SAIGE command for fitting null model
+    This will fit a Poisson / NB mixed model under the null hypothesis
+
+    Input:
+    - Phenotype / covariate file - rows: samples, cols: pheno (y), cov1, cov2 etc
+    - Comma separated str of column names from previous file to be used as covariates
+    - Same as above but for sample specific covariates
+    - Column name specifying sample / individual (e.g., IND_ID, or CPG_ID etc)
+    - output prefix: where to save the fitted model (.rda)
+    - Plink path: path to plink file (subset of ~2,000 markers for VRE)
+    - pheno col: name of column specifying pheno (default: "y")
+    - trait type: count = Poisson, count_nb = Negative Binomial, quantitative = Normal
+    - option to skip Variance Ratio estimation (discouraged)
+    - option to add an offset to the fixed covariates (???)
+    - option to transform (scale?) covariates?
+    - option to skip model fitting (discouraged)
+    - tolerance for convergence
+    - overwrite variance ratio file (estimated here)
+
+    Output:
+    Rscript command (str) ready to run (bash)
+    """
+    saige_command_step1 = 'Rscript step1_fitNULLGLMM_qtl.R'
+    saige_command_step1 += f' --useSparseGRMtoFitNULL={use_sparse_grm_null}'
+    saige_command_step1 += f' --useGRMtoFitNULL={use_grm_null}'
+    saige_command_step1 += f' --phenoFile={pheno_file}'
+    saige_command_step1 += f' --phenoCol={pheno_col}'
+    saige_command_step1 += f' --covarColList={cov_col_list}'
+    saige_command_step1 += f' --sampleCovarColList={sample_cov_col_list}'
+    saige_command_step1 += f' --sampleIDColinphenoFile={sample_id_pheno}'
+    saige_command_step1 += f' --traitType={trait_type}'
+    saige_command_step1 += f' --outputPrefix={output_prefix}'
+    saige_command_step1 += f' --skipVarianceRatioEstimation={skip_vre}'
+    saige_command_step1 += f' --isRemoveZerosinPheno={pheno_remove_zeros}'
+    saige_command_step1 += f' --isCovariateOffset={is_cov_offset}'
+    saige_command_step1 += f' --isCovariateTransform={is_cov_transform}'
+    saige_command_step1 += f' --skipModelFitting={skip_model_fitting}'
+    saige_command_step1 += f' --tol={tol}'
+    saige_command_step1 += f' --plinkFile={plink_path}'
+    saige_command_step1 += f' --IsOverwriteVarianceRatioFile={is_overwrite_vre_file}'
+    return saige_command_step1
```

Comment:
```suggestion
    return f"""
        Rscript step1_fitNULLGLMM_qtl.R \
        --useSparseGRMtoFitNULL={use_sparse_grm_null} \
        --useGRMtoFitNULL={use_grm_null} \
        --phenoFile={pheno_file} \
        --phenoCol={pheno_col} \
        --covarColList={cov_col_list} \
        --sampleCovarColList={sample_cov_col_list} \
        --sampleIDColinphenoFile={sample_id_pheno} \
        --traitType={trait_type} \
        --outputPrefix={output_prefix} \
        --skipVarianceRatioEstimation={skip_vre} \
        --isRemoveZerosinPheno={pheno_remove_zeros} \
        --isCovariateOffset={is_cov_offset} \
        --isCovariateTransform={is_cov_transform} \
        --skipModelFitting={skip_model_fitting} \
        --tol={tol} \
        --plinkFile={plink_path} \
        --IsOverwriteVarianceRatioFile={is_overwrite_vre_file}
    """
```

Your code here is great for optional arguments (e.g. only if `plink_path` is defined, append `--plinkFile={plink_path}`), but all your lines are inserted here so this may as well be a single block string

Reply:
would it be better to have this with only the needed files, and then append the optional ones only if defined? Any idea whether that makes any difference? Else what you did should be fine, as long as i make sure the default behaviour is defined for all pars

Reply:
You can definitely do a hybrid here, have one big block string with the mandatory arguments and then add optional arguments if specified 👍